ui/lib/src/nvui/handler.ts

```diff
@@ -125,24 +127,29 @@ export function selectionHandler(getOpponentColor: () => Color, isTouchDevice =
         // this is coupled to pieceJumpingHandler() noticing that the attribute is set and acting differently.
         if (rank === promotionRank && file && $firstPiece.attr('piece')?.toLowerCase() === 'p') {
           $evBtn.attr('promotion', 'true');
-          if (!isTouchDevice)
-            $boardLive.text('Promote to: q for queen, n for knight, r for rook, b for bishop');
-          else {
-            const queenPromotionKey = $(squareSelector(promotionRank === '8' ? '8' : '1', file));
-            const knightPromotionKey = $(squareSelector(promotionRank === '8' ? '7' : '2', file));
-            const rookPromotionKey = $(squareSelector(promotionRank === '8' ? '6' : '3', file));
-            const bishopPromotionKey = $(squareSelector(promotionRank === '8' ? '5' : '4', file));
-            const cancelPromotionKey = $(squareSelector(promotionRank === '8' ? '4' : '5', file));
-            queenPromotionKey.attr('promoteTo', 'q');
-            queenPromotionKey.text('promote to queen');
-            knightPromotionKey.attr('promoteTo', 'n');
-            knightPromotionKey.text('promote to knight');
-            rookPromotionKey.attr('promoteTo', 'r');
-            rookPromotionKey.text('promote to rook');
-            bishopPromotionKey.attr('promoteTo', 'b');
-            bishopPromotionKey.text('promote to bishop');
-            cancelPromotionKey.attr('promoteTo', 'x');
-            cancelPromotionKey.text('cancel');
+          if (!isTouchDevice) {
+            const msg = 'Promote to: q for queen, n for knight, r for rook, b for bishop';
+            $boardLive.text(msg + (isAntichess ? ', k for king' : ''));
+          } else {
+            let rank = promotionRank === '8' ? 8 : 1;
+            const promotionKeys = [
+              { role: 'q', text: 'promote to queen' },
+              { role: 'n', text: 'promote to knight' },
+              { role: 'r', text: 'promote to rook' },
+              { role: 'b', text: 'promote to bishop' },
+              { role: 'k', text: 'promote to king' },
+              { role: 'x', text: 'cancel' },
+            ];
+
+            for (const promotionKey of promotionKeys) {
+              if (promotionKey.role !== 'k' || isAntichess) {
+                const piecePromotionKey = $(squareSelector(rank.toString(), file));
+                piecePromotionKey.attr('promoteTo', promotionKey.role);
+                piecePromotionKey.text(promotionKey.text);
+              }
+              if (promotionRank === '8') rank--;
+              else rank++;
+            }
```

Comment:
```suggestion
            const promotions: { role: string; text: string }[] = [
              { role: 'q', text: 'promote to queen' },
              { role: 'n', text: 'promote to knight' },
              { role: 'r', text: 'promote to rook' },
              { role: 'b', text: 'promote to bishop' },
            ];
            if (isAntichess) promotions.push({ role: 'k', text: 'promote to king' });
            promotions.push({ role: 'x', text: 'cancel' });
            promotions.forEach(({ role, text }, index) => {
              const rank = promotionRank === '8' ? 8 - index : 1 + index;
              const piecePromotionEl = $(squareSelector(rank.toString(), file));
              piecePromotionEl.attr('promoteTo', role);
              piecePromotionEl.text(text);
            });
```

a bit clearer what's going on this way (to me, at least)